src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs

```diff
@@ -528,5 +541,116 @@ public void Dispose()
+            msbuildpath = GetMsBuildPathFromVSDirectory(vsDirectory, editions, targets);
+            if (!String.IsNullOrEmpty(msbuildpath)) return msbuildpath;
+
+            // 14.0, 12.0 
+            vsDirectory = new DirectoryInfo(Path.Combine(programFilesX86, "MSBuild"));
+            msbuildpath = GetMsBuildPathFromVSDirectoryBelow15(vsDirectory, editions, targets);
+            if (!String.IsNullOrEmpty(msbuildpath)) return msbuildpath;
+            return msbuildpath;
+        }
+
+        public static string GetMsBuildPathFromVSDirectory(DirectoryInfo vsDirectory, List<string> editions, string[] targets)
+        {
+            if (vsDirectory.Exists)
+            {
+                List<FileInfo> msBuildExePath = vsDirectory
+                    .GetDirectories("MSBuild", SearchOption.AllDirectories)
+                    .SelectMany(msBuildDir => msBuildDir.GetFiles("MSBuild.exe", SearchOption.AllDirectories))
+                    .OrderByDescending(msbuild => FileVersionInfo.GetVersionInfo(msbuild.FullName).FileVersion)
+                    .ThenBy(msbuild => editions.IndexOf(GetEditionType(msbuild.DirectoryName, editions)))
+                    .ThenBy(msbuild =>
+                    {
+                        var folderName = GetVersionFolder(msbuild.FullName);
+                        if (folderName.ToLower() == "current") return 1;
+                        else return -Convert.ToDouble(folderName);
+
+                    })
+                    .Where(msbuild =>
+                    {
+                        var targetsWithPath = GetTargetsWithPath(msbuild.DirectoryName, targets);
+                        if (targetsWithPath.TrueForAll(File.Exists)) return true;
+                        return false;
+                    })
+                    .ToList();
+                return msBuildExePath?.First().FullName;
+            };
+            return "";
+        }
+
+        public static string GetMsBuildPathFromVSDirectoryBelow15(DirectoryInfo vsDirectory, List<string> editions, string[] targets)
+        {
+            if (vsDirectory.Exists)
+            {
+                List<FileInfo> msBuildExePath = vsDirectory
+                    .GetFiles("MSBuild.exe", SearchOption.AllDirectories)
+                    .OrderByDescending(msbuild => FileVersionInfo.GetVersionInfo(msbuild.FullName).FileVersion)
+                    .ThenBy(msbuild =>
+                    {
+                        var folderName = GetVersionFolder(msbuild.FullName);
+                        if (folderName.ToLower() == "current") return 1;
+                        else return -Convert.ToDouble(folderName);
+
+                    })
+                    .Where(msbuild =>
+                    {
+                        var targetsWithPath = GetTargetsWithPath(msbuild.DirectoryName, targets);
+                        if (targetsWithPath.TrueForAll(File.Exists)) return true;
```

Comment:
Did you check if the same targets exist for lower versions?

Reply:
Yes they exist for MSBuild versions >= 14.0